### DISCUSSION TITLE
Should we keep a git hook for running code formatting?

### DISCUSSION THREAD

**Discussion by samhwang** [viet-aus-it/vait-discord-bot#43] [Ideas]
Tiếp nối cái discussion từ bên PR #23 . Có argument 2 hướng là ntn

Hướng 1:

- Tạo 1 cái git hook, pre-commit, trước khi anh em commit vào sẽ check linting.
- Tới đây có 2 hướng giải quyết:
  - Nó sẽ la làng lên khi anh em chưa format
  - HOẶC
  - Nó sẽ chạy 1 lệnh yarn format code hết cho anh em, xong nó sẽ commit.

Mục đích chính cho việc này là vì cứ push commit lên, xong cứ có 1 cái commit "Fix lint issue"
thì nhìn cái git log nó sẽ rất rối.

Hướng 2:

Không làm. Vì làm thì dev phải để cho ng ta hiểu tại sao lại phải lint. Vấn đề này nên là vấn đề từ trong tư tưởng,
chứ k phải nên rely vào tool.

Các anh em quyết định thế nào?

**Comment by samhwang** [comment]
@exNewbie @tklarryonline @hp0210

**Comment by dohaicuong** [comment]
em nghĩ là để lint sai lint thì nó hú để cho anh viết cho quen standards :))

**Comment by tklarryonline** [comment]
Hay là mình đi theo hướng 1 nhưng nó chỉ la làng lên cho anh chị em format, viết cho quen standards? :))

**Comment by samhwang** [comment]
> Hay là mình đi theo hướng 1 nhưng nó chỉ la làng lên cho anh chị em format, viết cho quen standards? :))

Tối hôm qua agree on đi theo hướng đó, tức là làm 1 cái hook, nó sẽ lint, nhưng k fix.

Cơ mà tối hôm qua em check lại thì có vẻ git pre-commit hook nó k push đc lên repo, phải set riêng từng người, k biết mọi ng chịu commit ko :( tạm thời em đang làm 1 cái là cài luôn cái lệnh cho mỗi lân start là nó tự test lint watch luôn.

**Comment by tklarryonline** [comment]
Em thử xem qua [husky](https://dev.to/devictoribero/how-to-use-husky-to-create-pre-commit-and-pre-push-hooks-4448) chưa?

**Comment by samhwang** [comment]
Em có check qua thử Husky, mà vấn đề là mình k có npm package ở folder gốc, mà chỉ có npm package ở 1 folder con. K lẽ phải chạy 2 lần `yarn install`, từ folder gốc vô folder con nữa...

**Comment by tklarryonline** [comment]
Well thì do có 2 projects trong 1 repo thôi. Em có thể xem thử https://typicode.github.io/husky/#/?id=custom-directory

**Comment by samhwang** [comment]
> Well thì do có 2 projects trong 1 repo thôi. Em có thể xem thử https://typicode.github.io/husky/#/?id=custom-directory

Vậy thì có hơi weird ko, khi 1 folder con nó lại in control of cái gì xảy ra ngoài folder gốc...

**Comment by tklarryonline** [comment]
1 repo 2 Projects con với 2 package.json files thì cũng weird rồi.

**Comment by samhwang** [comment]
Yeah nên em k muốn nó weird hơn :-/ Nếu làm theo hướng anh nói có khi phải restructure lại thành lerna-style monorepo, r xong build docker theo hướng khác.

Mà cái đó lại phải nghiên cứu cái Lerna nó ntn... and then nó rối như tơ nhện...

**Comment by samhwang** [comment]
https://dev.to/krzysztofkaczy9/do-you-really-need-husky-247b

hmm...

**Comment by tklarryonline** [comment]
Anh nghĩ là em đang đi lòng vòng nếu muốn đi hướng 1. Vì kiểu gì cũng phải có 1 cách nào đó cài git-hook vào trước khi commit & push. 1 là 2-step installer hoặc 2 là cái sub-project enforce cho cái tổng bên ngoài.

**Comment by samhwang** [comment]
Note: Đã add git-hook trong PR #20
